libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -0,0 +1,801 @@
+/*
+ * AP_UAVCAN.cpp
+ *
+ *      Author: Eugene Shamaev
+ */
+
+#include <AP_Common/AP_Common.h>
+#include <AP_HAL/AP_HAL.h>
+
+#if HAL_WITH_UAVCAN
+
+#include "AP_UAVCAN.h"
+#include <GCS_MAVLink/GCS.h>
+
+// Zubax GPS and other GPS, baro, magnetic sensors
+#include <uavcan/equipment/gnss/Fix.hpp>
+#include <uavcan/equipment/gnss/Auxiliary.hpp>
+#include <uavcan/equipment/ahrs/MagneticFieldStrength.hpp>
+#include <uavcan/equipment/air_data/StaticPressure.hpp>
+#include <uavcan/equipment/air_data/StaticTemperature.hpp>
+#include <uavcan/equipment/actuator/ArrayCommand.hpp>
+#include <uavcan/equipment/actuator/Command.hpp>
+#include <uavcan/equipment/actuator/Status.hpp>
+#include <uavcan/equipment/esc/RawCommand.hpp>
+
+#include <AP_BoardConfig/AP_BoardConfig.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define debug_uavcan(level, fmt, args...) do { if ((level) <= AP_BoardConfig::get_can_debug()) { hal.console->printf(fmt, ##args); }} while (0)
+
+// Translation of all messages from UAVCAN structures into AP structures is done
+// in AP_UAVCAN and not in corresponding drivers.
+// The overhead of including definitions of DSDL is very high and it is best to
+// concentrate in one place.
```

Comment:
@pavel-kirienko That's some ugly code to look at :smile: Any chance this will change in the future?

I guess we don't have an option here, but this drivers-in-another-library won't be pretty to see.